include/sparrow/allocator.hpp

```diff
@@ -0,0 +1,247 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <memory_resource>
+#include <type_traits>
+#include <variant>
+
+namespace sparrow
+{
+    /*
+     * allocator concept, based on the requirements specified by the standard:
+     * https://en.cppreference.com/w/cpp/named_req/Allocator.
+     */
+    template <class T>
+    concept allocator = std::copy_constructible<std::remove_cvref_t<T>>
+        and std::move_constructible<std::remove_cvref_t<T>>
+        and std::equality_comparable<std::remove_cvref_t<T>>
+        and requires(std::remove_cvref_t<T>& alloc,
+                     typename std::remove_cvref_t<T>::value_type* p,
+                     std::size_t n)
+        {
+            typename std::remove_cvref_t<T>::value_type;
+            { alloc.allocate(n) } -> std::same_as<typename std::remove_cvref_t<T>::value_type*>;
+            { alloc.deallocate(p, n) } -> std::same_as<void>;
+        };
+
+    /*
+     * Type erasure class for allocators. This allows to use any kind of allocator
+     * (standard, polymorphic) without having to expose it as a template parameter.
+     *
+     * @tparam T value_type of the allocator
+     * @tparam DA default allocator, instantiated when calling the default constructor
+     */
+    template <class T, class DA = std::allocator<T>>
+    class any_allocator
+    {
+        template <class A>
+        static constexpr bool is_value_stored = 
+            std::is_same_v<A, std::allocator<T>> || std::is_same_v<A, std::pmr::polymorphic_allocator<T>>;
+
+    public:
+
+        using value_type = T;
+
+        any_allocator();
+        any_allocator(const any_allocator& rhs);
+        any_allocator(any_allocator&&) = default;
+
+        any_allocator& operator=(const any_allocator& rhs) = delete;
+        any_allocator& operator=(any_allocator&& rhs) = delete;
+
+        template <class A>
+        any_allocator(A&& alloc)
+        requires (not std::same_as<std::remove_cvref_t<A>, any_allocator>
+                and sparrow::allocator<A>)
+            : m_storage(make_storage(std::forward<A>(alloc)))
+        {
+        }
+
+        [[nodiscard]] T* allocate(std::size_t n);
+        void deallocate(T* p, std::size_t n);
+        
+        any_allocator select_on_container_copy_construction() const;
+
+        bool equal(const any_allocator& rhs) const; 
+
```

Comment:
Consider adding
```c++
contexpr std::type_index typeid() const;
```
So that the type can be introspectable in a cheap way.

Reply:
I'm not sure we need it as a public API. What the `any_allocator` type really holds should remain private in my opinion.

Reply:
Wouldn't user code that need to transfer the data at some point need to know that information? not sure.